internal/config/config.go

```diff
@@ -254,6 +274,7 @@ func Parse(bs []byte) (*Config, error) {
 }
 
 func parsePeer(p peer) (*Peer, error) {
+	_, frrMode := os.LookupEnv("METALLB_BGP_TYPE")
```

Comment:
I'd avoid checking the env variable here, it's gonna be replaced by a flag (see https://github.com/metallb/metallb/pull/958#discussion_r723083723).

I think the best way would be to keep the config unaware of the implementation, and fail in the speakers `SetConfig` method: https://github.com/metallb/metallb/blob/9cb1653a87673ec91971a60c873204e0752ff97b/speaker/main.go#L347

specifically in the bgp handler: https://github.com/metallb/metallb/blob/9cb1653a87673ec91971a60c873204e0752ff97b/speaker/bgp_controller.go#L58

That would cause the parsing to fail, but the failure will happen in the frr-aware part (or un-aware) part. That would also require the default not to be set here but when we parse the config, in the bgpfrr side.

Reply:
this is kiind of pretty early way b4 any parsing is done ? if we need this to be done in implementation independent way we may have to create interface for parsePeer() parseCommity()  parseAddresspool() only to handle the delta and keep the rest of the parsing common as is but that means the too stack will live together my understanding at somepoint we will just default to FRR so we need to be careful how much we need to invest on this atm we need only two additional knobs if that evolved we can look at more fancy approach not sure

Reply:
I am just saying to move the check and the defaulting place to another function.
I'd say let's fill the config with all the parameters the user put in place, and defer to where we consume it to choose if it's good or not.
The parsing + consuming happens in the same stack of calls so it won't be a big deal from a UX point of view.